File: easysql_web/src/api/fewShot.ts
import { apiClient } from './client';

export interface FewShotCreateRequest {
  db_name: string;
  question: string;
  sql: string;
  tables_used?: string[];
  explanation?: string;
  message_id?: string;
}

export interface FewShotUpdateRequest {
  question?: string;
  sql?: string;
  tables_used?: string[];
  explanation?: string;
}

export interface FewShotInfo {
  id: string;
  db_name: string;
  question: string;
  sql: string;
  tables_used: string[];
  explanation: string | null;
  message_id: string | null;
  created_at: string;
}

export interface FewShotListResponse {
  items: FewShotInfo[];
  total: number;
  db_name: string;
}

export interface FewShotCheckResponse {
  is_few_shot: boolean;
  example_id: string | null;
  example: FewShotInfo | null;
}

export interface DuplicateErrorDetail {
  message: string;
  existing_id: string;
  similarity_score: number;
}

export const fewShotApi = {
  create: async (data: FewShotCreateRequest): Promise<FewShotInfo> => {
    const response = await apiClient.post<FewShotInfo>('/few-shot', data);
    return response.data;
  },

  list: async (
    dbName: string,
    limit = 100,
    offset = 0
  ): Promise<FewShotListResponse> => {
    const response = await apiClient.get<FewShotListResponse>('/few-shot', {
      params: { db_name: dbName, limit, offset },
    });
    return response.data;
  },

  getById: async (id: string): Promise<FewShotInfo> => {
    const response = await apiClient.get<FewShotInfo>(`/few-shot/${id}`);
    return response.data;
  },

  update: async (id: string, data: FewShotUpdateRequest): Promise<FewShotInfo> => {
    const response = await apiClient.put<FewShotInfo>(`/few-shot/${id}`, data);
    return response.data;
  },

  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/few-shot/${id}`);
  },

  checkByMessageId: async (messageId: string): Promise<FewShotCheckResponse> => {
    const response = await apiClient.get<FewShotCheckResponse>(
      `/few-shot/by-message/${messageId}`
    );
    return response.data;
  },
};

File: easysql_api/models/few_shot.py
"""Few-Shot API data models."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class FewShotCreate(BaseModel):
    """Request to create a new few-shot example."""

    db_name: str = Field(..., description="Database name for isolation")
    question: str = Field(..., description="Natural language question")
    sql: str = Field(..., description="SQL query")
    tables_used: list[str] = Field(default_factory=list, description="Tables used in query")
    explanation: str | None = Field(default=None, description="Optional explanation")
    message_id: str | None = Field(default=None, description="Related message ID")


class FewShotUpdate(BaseModel):
    """Request to update an existing few-shot example."""

    question: str | None = Field(default=None, description="Updated question")
    sql: str | None = Field(default=None, description="Updated SQL query")
    tables_used: list[str] | None = Field(default=None, description="Updated tables list")
    explanation: str | None = Field(default=None, description="Updated explanation")


class FewShotInfo(BaseModel):
    """Few-shot example information."""

    id: str
    db_name: str
    question: str
    sql: str
    tables_used: list[str]
    explanation: str | None
    message_id: str | None
    created_at: datetime


class FewShotList(BaseModel):
    """List of few-shot examples."""

    items: list[FewShotInfo]
    total: int
    db_name: str


class FewShotCheckResponse(BaseModel):
    """Response for checking if a message is saved as few-shot."""

    is_few_shot: bool
    example_id: str | None = None
    example: FewShotInfo | None = None
